server/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -1,90 +1,125 @@
-          })
+  trait TokenBucket[F[_]] {
+    def takeToken: F[TokenAvailability]
+  }
+
+  object TokenBucket {
+
+    /**
+      * Creates an in-memory [[TokenBucket]].
+      *
+      * @param capacity the number of tokens the bucket can hold and starts with.
+      * @param refillEvery the frequency with which to add another token if there is capacity spare.
+      * @return A task to create the [[TokenBucket]].
+      */
+    def local[F[_]](capacity: Int, refillEvery: FiniteDuration)(
+        implicit F: Sync[F],
+        clock: Clock[F]): F[TokenBucket[F]] = {
+
+      def getTime = clock.monotonic(NANOSECONDS)
+      val bucket = getTime.flatMap(time => Ref[F].of((capacity.toDouble, time)))
+
+      bucket.map { counter =>
+        new TokenBucket[F] {
+          override def takeToken: F[TokenAvailability] = {
+            val attemptUpdate = counter.access.flatMap {
+              case ((previousTokens, previousTime), setter) =>
+                getTime.flatMap(currentTime => {
+                  val timeDifference = currentTime - previousTime
+                  val tokensToAdd = timeDifference.toDouble / refillEvery.toNanos.toDouble
+                  val newTokenTotal = Math.min(previousTokens + tokensToAdd, capacity.toDouble)
+
+                  val attemptSet: F[Option[TokenAvailability]] = if (newTokenTotal >= 1) {
+                    setter((newTokenTotal - 1, currentTime))
+                      .map(_.guard[Option].as(TokenAvailable()))
+                  } else {
+                    val timeToFull = (refillEvery.toNanos * capacity) - timeDifference
+                    val successResponse = TokenUnavailable(timeToFull.nanos.some)
+                    setter((newTokenTotal, currentTime)).map(_.guard[Option].as(successResponse))
+                  }
+
+                  attemptSet
+                })
+            }
+
+            def loop: F[TokenAvailability] = attemptUpdate.flatMap { attempt =>
+              attempt.fold(loop)(token => token.pure[F])
+            }
+            loop
+          }
+
+        }
       }
-      Stream(bucket).concurrently(refill)
+
     }
-}
 
-/**
-  * Transform a service to reject any calls the go over a given rate with a 429 response.
-  */
-object Throttle {
+  }
 
   /**
     * Limits the supplied service to a given rate of calls using an in-memory [[TokenBucket]]
     *
     * @param amount the number of calls to the service to permit within the given time period.
     * @param per the time period over which a given number of calls is permitted.
     * @param http the service to transform.
-    * @return a singleton stream containing the transformed service.
+    * @return a task containing the transformed service.
     */
   def apply[F[_], G[_]](amount: Int, per: FiniteDuration)(
-      http: Http[F, G])(implicit F: Concurrent[F], timer: Timer[F]): Stream[F, Http[F, G]] = {
+      http: Http[F, G])(implicit F: Sync[F], timer: Clock[F]): F[Http[F, G]] = {
```

Comment:
This is kind of confusing: the types are the same as `TokenBucket.local`, but we do extra math first.

I wonder whether this overload carries its weight at all, given that it doesn't allow customization of the throttled response.  I imagine we could live without it, but I'd defer to you since you've probably tried to use this and I haven't.

Reply:
I'm not totally attached but my thinking was that some people will not care about the token bucket impl so easier to have a default where you don't need to construct your own, and the x requests per y time was a more "natural" way that people would think about their rate limit, ie "100 per second" rather than "1 per 0.6 seconds"

Reply:
I'm very slightly 👎 on the overload